packages/ampd-sdk/src/grpc/connection_pool.rs

```diff
@@ -0,0 +1,394 @@
+                        .await?;
+
+                        let mut receiver = self.connection_receiver.clone();
+                        receiver.changed().await.into_report()?;
+                    }
+                    ConnectionState::Reconnecting => {
+                        let mut receiver = self.connection_receiver.clone();
+                        receiver.changed().await.into_report()?;
+                    }
+                }
+            }
+        })
+        .await
+        .into_report()?
+    }
+
+    pub async fn send_client_message(&self, message: ClientMessage) -> Result<(), Error> {
+        self.message_sender.send(message).await.into_report()
+    }
+
+    pub fn connection_receiver(&self) -> watch::Receiver<ConnectionState> {
+        self.connection_receiver.clone()
+    }
+}
+
+/// Connection pool that handles connection state and retries.
+///
+/// This is a multi-producer single consumer pattern:
+/// - The pool maintains the connection state via a watch channel
+/// - Multiple clients can subscribe to connection state changes
+/// - Clients communicate back to the pool via mpsc channel
+/// - The pool handles connection retries and failure recovery
+#[derive(Debug)]
+pub struct ConnectionPool {
+    url: Url,
+    connection_state: watch::Sender<ConnectionState>,
+    message_receiver: mpsc::Receiver<ClientMessage>,
+    retry_policy: RetryPolicy,
+    is_reconnecting: bool,
+}
+
+impl ConnectionPool {
+    pub fn new(url: Url) -> (Self, ConnectionHandle) {
+        let (state_sender, state_receiver) = watch::channel(ConnectionState::Disconnected);
+        let (msg_sender, msg_receiver) = mpsc::channel(MESSAGE_RECEIVER_CHANNEL_CAPACITY);
+
+        let handle = ConnectionHandle {
+            connection_receiver: state_receiver,
+            message_sender: msg_sender,
+        };
+
+        let pool = Self {
+            url,
+            connection_state: state_sender,
+            message_receiver: msg_receiver,
+            retry_policy: DEFAULT_RETRY_POLICY,
+            is_reconnecting: false,
+        };
+
+        (pool, handle)
+    }
+
+    pub async fn run(mut self, token: CancellationToken) -> Result<(), Error> {
+        if let Err(e) = self.connect().await {
+            warn!(err = ?e, "initial connection failed, will retry on client requests");
+        }
+
+        loop {
+            tokio::select! {
+                message = self.message_receiver.recv() => {
+                    match message {
+                        Some(ClientMessage::ConnectionFailed(details)) => {
+                            warn!("client reported connection failure: {}", details);
+                            if !self.is_reconnecting {
+                                if let Err(e) = self.handle_connection_failure().await {
+                                    warn!(err = ?e, "reconnection attempt failed, will retry on next client request");
+                                }
+                            }
```

Comment:
there is an issue here: let's say we have 10 clients. All of them fail to make a request call and notify the pool. The first message will reconnect and wait until that's done. After that we have a newly established connection, but still 9 failure notifications in the channel. That means the pool will kill and reconnect the connection another 9 times.
It needs to be something like this:
- receive failure notification
- set status to reconnecting
- drain failure notification channel
- actually reconnect while continuing to drain (`select!`)
- when connection is established change the state and stop draining